### ISSUE TITLE
Merge MultiApiTestRunner functionality into RobolectricTestRunner

### ISSUE THREAD

**Issue by jongerrish** [robolectric/robolectric#2438]
Ability to run tests on multiple apis should be built into RobolectricTestRunner.

Make @Config.sdk @Deprecated add @Config.minSdk @Config.maxSdk - since sdk is an int array, specifying ranges is messy (can't use constants such as LOLLIPOP_AND_ABOVE = {21, 22, 23} which would also need to be maintained by hand).

maxSdk should be optional - omitting it should use max robolectric supported api
minSdk should be optional - omitting it should use min robolectric supported api.

**Comment by fiower** [comment]
What if someone wants to run test only at API 21 and 23 but not 22 (just an example)?

**Comment by jongerrish** [comment]
Do you have a specific case in mind?

Android featured are usually added at a particular api level (and
occasionally removed at another). I can't think of a case where you'd want
to test non-contiguous api levels.

**Comment by fiower** [comment]
I don't have specific case. But just thinking it might be possible. For example, if I know some milestone changes happened in API 19 and then API 23, and I only want to test against these levels. 

But yes, the most common use case are contiguous api levels.

**Comment by karlicoss** [comment]
Not sure if I fully understood the proposal, would it still be possible to run tests on specific (non-contiguous) API if one specifies them in config directly (by injecting in the app-specific test runner, or via `robolectric.properties`)? If yes, then I guess it's fine; if not, it might be a problem for people who've got too much tests to run against all of APIs and just run on min and target sdks.

**Comment by karlicoss** [comment]
@jongerrish as for "can't use constants such as LOLLIPOP_AND_ABOVE = {21, 22, 23}", I think, people usually handle that by specfying arrays as Strings (you are allowed to reference constant strings in annotations), although it's certainly not very convenient and still leaves the issue of maintaining these constants. But, that way it's at least kinda flexible.

**Comment by jongerrish** [comment]
We to make is such that running tests on all apis is not prohibitively slow but maybe there are ways to improve this:-

add minSdk + maxSdk to run ranges
keep sdk = {} for running specific non-contiguous api versions
add some kind of property / control for sdkTypes = { ALL, MIN_AND_TARGET, TARGET_ONLY } so projects could set up say a smoke build which runs TARGET_ONLY and a more complete build that runs ALL. Not sure this should be tied to the code though, probably some system property the build could set?

**Comment by kingargyle** [comment]
I've written tests to verify that specific dimension values are being used based on different APIS, and also that different code is being injected and returned based on the API.   Need a way at least at the @Test method level to still specifiy what API Level the test should be run against.

**Comment by jongerrish** [comment]
Would @Config.minSdk / @Config.maxSdk not support that also?

**Comment by kingargyle** [comment]
Would I need to do:

@Test @Config(minSdk = 18, maxSdk = 18) 

to just test sdk 18?

I can see the value for a range of apis as well.

One thing to consider is leveraging the junit Categories annotation.

https://github.com/junit-team/junit4/wiki/Categories

Could specify APIs in particular categories.

**Comment by xian** [comment]
The plan at the moment is to make @Config(sdk=…) and @Config(minSdk=…,maxSdk=…) mutually exclusive at any given level of configuration (e.g. per test method, class, or properties file), and mutually overriding between levels.

For example:

```java
@RunWith(RobolectricTestRunner.class) @Config(minSdk=21)
class ATest {
  public void testMethod0() { … } // runs on 21 through highest supported
  @Config(maxSdk=22) public void testMethod1() { … } // runs on 21 and 22
  @Config(sdk=16) public void testMethod2() { … } // runs on 16
}
```

It's unclear to me how `AndroidManifest.xml`'s minSdkVersion should interact. Thoughts?

**Comment by xian** [comment]
Underway at #2772.

**Comment by xian** [comment]
Merged, will be in 3.2. Docs at http://robolectric.org/configuring/.
